state_machine/applications/flight/lib/orbital_mechanics.py

```diff
@@ -0,0 +1,78 @@
+# Based on: https://github.com/RoboticExplorationLab/pycubed_circuitpython/blob/master/Scheduler/sched2.py
+try:
+    from ulab.numpy import zeros, array
+    from ulab.numpy.linalg import norm
+except ImportError:
+    from numpy import zeros, array
+    from numpy.linalg import norm
+
+MU = 3.986004418E5
+R = 6371.009  # Earth's radius (km)
+J2 = 1.08262668E-3  # J2 Value
+
+def rk4(x, dt, f):
+    """Runge-Kutta order 4 integration
+    Args:
+        - x: 6-vector, first 3 are position, last 3 are velocity
+        - dt: time step
+        - f: function defining the time derivative of x, ie: f(x)=dx
+
+    Returns:
+        - x: The integrated state
+    """
+    k1 = dt * f(x)
+    k2 = dt * f(x + k1 / 2)
+    k3 = dt * f(x + k2 / 2)
+    k4 = dt * f(x + k3)
+    return x + (1 / 6) * (k1 + 2 * k2 + 2 * k3 + k4)
+
+def d_state(x):
+    """Time derivative of the state variable
+    Args:
+        - x: 6-vector, first 3 are position (km), last 3 are velocity (km/s)
```

Comment:
kinda weird to have the state units be in km. Is there a specific reason you chose this?

Reply:
We are dealing with **really** large numbers, we should get better numerical precision with km instead of m. Also I believe down the chain we feed this into the IGRF13 model which uses km anyway (probably for the same reasons).